csidriver/pkg/app/controllerserver.go

```diff
@@ -0,0 +1,453 @@
+		volumeID)
+	if err != nil {
+		klog.Errorf("build message resource failed with error: %s", err)
+		return nil, err
+	}
+
+	m := jsonpb.Marshaler{}
+	js, err := m.MarshalToString(req)
+	if err != nil {
+		klog.Errorf("failed to marshal to string with error: %s", err)
+		return nil, err
+	}
+	klog.Infof("controller publish volume marshal to string: %s", js)
+	msg.Content = js
+	msg.BuildRouter(DefaultReceiveModuleName,
+		GroupResource,
+		resource,
+		constants.CSIOperationTypeControllerPublishVolume)
+
+	// Marshal message
+	reqData, err := json.Marshal(msg)
+	if err != nil {
+		klog.Errorf("marshal request failed with error: %v", err)
+		return nil, err
+	}
+
+	// Send message to KubeEdge
+	resdata, err := send2KubeEdge(string(reqData), cs.keEndpoint)
+	if err != nil {
+		klog.Errorf("send to kubeedge failed with error: %v", err)
+		return nil, err
+	}
+
+	// Unmarshal message
+	result, err := extractMessage(resdata)
+	if err != nil {
+		klog.Errorf("unmarshal response failed with error: %v", err)
+		return nil, err
+	}
+
+	klog.Infof("controller publish volume result: %v", result)
+	data := result.GetContent().(string)
+
+	if msg.GetOperation() == model.ResponseErrorOperation {
+		klog.Errorf("controller publish volume with error: %s", data)
+		return nil, errors.New(data)
+	}
+
+	decodeBytes, err := base64.StdEncoding.DecodeString(data)
+	if err != nil {
+		klog.Errorf("controller publish volume decode with error: %v", err)
+		return nil, err
+	}
+
+	controllerPublishVolumeResponse := &csi.ControllerPublishVolumeResponse{}
+	err = json.Unmarshal([]byte(decodeBytes), controllerPublishVolumeResponse)
+	if err != nil {
+		klog.Errorf("controller publish volume unmarshal with error: %v", err)
+		return nil, nil
+	}
+	klog.Infof("controller publish volume response: %v", controllerPublishVolumeResponse)
+	return controllerPublishVolumeResponse, nil
+}
+
+// ControllerUnpublishVolume issues controller unpublish volume func
+func (cs *controllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	instanceID := req.GetNodeId()
+	volumeID := req.GetVolumeId()
+
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerUnpublishVolume Volume ID must be provided")
+	}
+
+	if len(instanceID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerUnpublishVolume Instance ID must be provided")
+	}
+
+	// Build message struct
+	msg := model.NewMessage("")
+	resource, err := buildResource(cs.nodeID,
+		DefaultNamespace,
+		constants.CSIResourceTypeVolume,
+		volumeID)
+	if err != nil {
+		klog.Errorf("Build message resource failed with error: %s", err)
+		return nil, err
+	}
+
+	m := jsonpb.Marshaler{}
+	js, err := m.MarshalToString(req)
+	if err != nil {
+		klog.Errorf("failed to marshal to string with error: %s", err)
+		return nil, err
+	}
+	klog.Infof("controller Unpublish Volume marshal to string: %s", js)
```

Comment:
can we log this with V.(3) or V.(4), too?

Reply:
I will. Thanks